pkg/typesystem/typesystem_create_test.go

```diff
@@ -0,0 +1,323 @@
+	case *openfgav1.Userset_TupleToUserset:
+		y, ok := usersetB.(*openfgav1.Userset_TupleToUserset)
+		require.True(t, ok)
+		tupleToUsersetEquals(t, x.TupleToUserset, y.TupleToUserset)
+	case *openfgav1.Userset_Union:
+		y, ok := usersetB.(*openfgav1.Userset_Union)
+		require.True(t, ok)
+		usersetsEquals(t, x.Union, y.Union)
+	case *openfgav1.Userset_Intersection:
+		y, ok := usersetB.(*openfgav1.Userset_Intersection)
+		require.True(t, ok)
+		usersetsEquals(t, x.Intersection, y.Intersection)
+	case *openfgav1.Userset_Difference:
+		y, ok := usersetB.(*openfgav1.Userset_Difference)
+		require.True(t, ok)
+		if x.Difference == nil || y.Difference == nil {
+			require.Equal(t, x.Difference, y.Difference)
+			return
+		}
+		rewriteEquals(t, x.Difference.GetBase(), y.Difference.GetBase())
+		rewriteEquals(t, x.Difference.GetSubtract(), y.Difference.GetSubtract())
+	default:
+		require.Equal(t, usersetA, usersetB)
+	}
+}
+
+func relationReferenceEquals(t *testing.T, a, b *openfgav1.RelationReference) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetCondition(), b.GetCondition())
+	require.Equal(t, a.GetType(), b.GetType())
+
+	rwA := a.GetRelationOrWildcard()
+	rwB := b.GetRelationOrWildcard()
+
+	switch x := rwA.(type) {
+	case *openfgav1.RelationReference_Wildcard:
+		y, ok := rwB.(*openfgav1.RelationReference_Wildcard)
+		require.True(t, ok)
+		require.Equal(t, x.Wildcard, y.Wildcard)
+	case *openfgav1.RelationReference_Relation:
+		y, ok := rwB.(*openfgav1.RelationReference_Relation)
+		require.True(t, ok)
+		require.Equal(t, x.Relation, y.Relation)
+	default:
+		require.Equal(t, rwA, rwB)
+	}
+}
+
+func relationTypeInfoEquals(t *testing.T, a, b *openfgav1.RelationTypeInfo) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	directA := a.GetDirectlyRelatedUserTypes()
+	directB := b.GetDirectlyRelatedUserTypes()
+
+	if directA == nil || directB == nil {
+		require.Equal(t, directA, directB)
+		return
+	}
+	require.Equal(t, len(directA), len(directB))
+	for i := 0; i < len(directA); i++ {
+		relationReferenceEquals(t, directA[i], directB[i])
+	}
+}
+
+func relationEquals(t *testing.T, a, b *openfgav1.Relation) {
+	require.Equal(t, a.GetName(), b.GetName())
+	rewriteEquals(t, a.GetRewrite(), b.GetRewrite())
+	relationTypeInfoEquals(t, a.GetTypeInfo(), b.GetTypeInfo())
+}
+
+func sourceInfoEquals(t *testing.T, a, b *openfgav1.SourceInfo) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetFile(), b.GetFile())
+}
+
+func relationMetadataEquals(t *testing.T, a, b *openfgav1.RelationMetadata) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetModule(), b.GetModule())
+	sourceInfoEquals(t, a.GetSourceInfo(), b.GetSourceInfo())
+
+	directA := a.GetDirectlyRelatedUserTypes()
+	directB := b.GetDirectlyRelatedUserTypes()
+	if directA == nil || directB == nil {
```

Comment:
Similar to comment above, consider asserting directA and directB to be non nil

Reply:
`GetDirectlyRelatedUserTypes` returns a slice, so it most certainly can be `nil`.